src/backend/aspen/cli/sync_auth0.py

```diff
@@ -0,0 +1,420 @@
+#!/usr/bin/env python3
+import logging
+import os
+import random
+import string
+from functools import cache, partial
+from typing import Any, MutableSequence, Tuple, TypedDict
+
+import click
+import sqlalchemy as sa
+from auth0.v3 import authentication as auth0_authentication
+from auth0.v3.management import Auth0
+
+from aspen.api.settings import Settings
+from aspen.config.config import Config
+from aspen.database.connection import (
+    get_db_uri,
+    init_db,
+    session_scope,
+    SqlAlchemyInterface,
+)
+from aspen.database.models import Group, User
+
+
+class Auth0Org(TypedDict):
+    id: str
+    name: str
+    display_name: str
+
+
+class Auth0User(TypedDict):
+    id: str
+    name: str
+
+
+def generate_password(length=22):
+    possible_characters = (
+        string.ascii_uppercase + string.ascii_lowercase + string.digits
+    )
+    return "".join(random.choice(possible_characters) for _ in range(length))
+
+
+class Auth0Client:
+    def __init__(self):
+        # TODO these will need to be read from settings instead of env.
+        client_id = os.environ.get("AUTH0_CLIENT_ID")
+        client_secret = os.environ.get("AUTH0_CLIENT_SECRET")
+        domain = os.environ.get("AUTH0_DOMAIN")
+        auth_req = auth0_authentication.GetToken(domain)
+        token = auth_req.client_credentials(
+            client_id, client_secret, f"https://{domain}/api/v2/"
+        )
+
+        self.client = Auth0(domain=domain, token=token["access_token"])
+
+    def get_all_results(self, endpoint, key):
+        # Auth0 paginates results. We don't have a crazy amount of data in auth0 so we can
+        # afford to just paginate through all the results and hold everything in memory.
```

Comment:
at what scale would this stop being true?

Reply:
As far as memory concerns go, we can handle well into the tens, or even low hundreds of thousands of objects at a time, but I think it's likely that we'll get grouchy earlier than that about all of the up-front latency for making so many API calls to Auth0 before we even start processing. I think we'll might want to revisit the structure of this script in the ~5k-10k users/groups range.